src/java/org/apache/cassandra/utils/concurrent/NonBlockingRateLimiter.java

```diff
@@ -0,0 +1,183 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.utils.concurrent;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Ticker;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A rate limiter implementation that allows callers to reserve permits that may only be available 
+ * in the future, delegating to them decisions about how to schedule/delay work and whether or not
+ * to block execution to do so.
+ */
+@SuppressWarnings("UnstableApiUsage")
+@ThreadSafe
+public class NonBlockingRateLimiter
+{
+    private static final long DEFAULT_BURST_NANOS = TimeUnit.SECONDS.toNanos(1);
+    
+    /** a starting time for elapsed time calculations */
+    private final long startedNanos;
+    
+    /** 
+     * nanoseconds of "burst time"
+     * 
+     * ex. If this is configured at the default of 1 second, the limiter will allow a
+     *     number of permits equal to the configured number of permits/second to be issued 
+     *     in a "burst" prior to the limiter reaching a steady state. This allows the limiter 
+     *     to adhere more closely to the configured rate, as it can bring forward short
+     *     periods of inactivity if the rate of requests oscillates around the configured limit.
+     */
+    private final long burstNanos;
+
+    /** nanoseconds from start time corresponding to the next available permit */
+    private final AtomicLong nextAvailable = new AtomicLong();
+    
+    private volatile Ticker ticker;
+    
+    private volatile int permitsPerSecond;
+    
+    /** time in nanoseconds between permits on the timeline */
+    private volatile long intervalNanos;
+
+    public NonBlockingRateLimiter(int permitsPerSecond)
+    {
+        this(permitsPerSecond, DEFAULT_BURST_NANOS, Ticker.systemTicker());
+    }
+
+    @VisibleForTesting
+    public NonBlockingRateLimiter(int permitsPerSecond, long burstNanos, Ticker ticker)
+    {
+        this.startedNanos = ticker.read();
+        this.burstNanos = burstNanos;
+        setRate(permitsPerSecond, ticker);
+    }
+
+    public void setRate(int permitsPerSecond)
+    {
+        setRate(permitsPerSecond, Ticker.systemTicker());
+    }
+
+    @VisibleForTesting
+    public synchronized void setRate(int permitsPerSecond, Ticker ticker)
+    {
+        Preconditions.checkArgument(permitsPerSecond > 0, "permits/second must be positive");
+
+        this.ticker = ticker;
+        this.permitsPerSecond = permitsPerSecond;
+        intervalNanos = Math.max(TimeUnit.SECONDS.toNanos(1) / permitsPerSecond, 1);
```

Comment:
If someone provides < 1 permit per second and we override it, we should probably log a warning to let someone know that's happened.

Reply:
I think it's basically impossible for this to happen anyway, given `permitsPerSecond` has to be at least 1. Any objections to just removing the `max` expression? The only other thing I might do is set a limit of `TimeUnit.SECONDS.toNanos(1)` on `permitsPerSecond`.

Reply:
Nah, that's fine. I didn't even think about us validating the range of the param further upstream from here.